File: AirHockeyLightClient/AirHockeyLightClient/Noyau/Arbre/Visiteurs/VisiteurDeplacement.swift
///////////////////////////////////////////////////////////////////////////////
/// @file VisiteurDeplacement.swift
/// @author Pierre To
/// @date 2017-10-10
/// @version 1
///
/// @addtogroup log3900 LOG3990
/// @{
///////////////////////////////////////////////////////////////////////////////

import GLKit
import SceneKit

///////////////////////////////////////////////////////////////////////////
/// @class VisiteurDeplacement
/// @brief Permet de deplacer les noeuds
///
/// @author Pierre To
/// @date 2017-10-10
///////////////////////////////////////////////////////////////////////////
class VisiteurDeplacement: VisiteurAbstrait {
    
    /// Vecteur trois dimensions pour le changement de position de l'objet
    private var delta: GLKVector3?
    
    /// Constructeur
    init(delta: GLKVector3) {
        self.delta = delta
    }
    
    /// Visiter un accélérateur pour le déplacement
    func visiterAccelerateur(noeud: NoeudAccelerateur) {
        if (noeud.estSelectionne()) {
            self.deplacerNoeud(noeud: noeud)
        }
    }
    
    /// Visiter un maillet pour le déplacement
    //virtual void visiterMaillet(NoeudMaillet* noeud);
    
    /// Visiter une table pour le déplacement
    func visiterTable(noeud: NoeudTable) {
    }
    
    /// Visiter un point de contrôle pour le déplacement
    func visiterPointControl(noeud: NoeudPointControl) {
        if (noeud.estSelectionne()) {
            var pos = noeud.obtenirPositionRelative()
            pos = GLKVector3Add(pos, self.delta!)
            
            // Deplacer le premier noeud
            noeud.deplacer(position: pos)
            
            // Deplacer son opposé
            let posNoeudOppose = GLKVector3Multiply(pos, noeud.obtenirSymmetrie())
            noeud.obtenirNoeudOppose().deplacer(position: posNoeudOppose)
            
            // Envoyer la commande
            FacadeModele.instance.obtenirEtatEdition().currentUserControlPointChanged(uuid: noeud.obtenirUUID(), pos: noeud.position)
            FacadeModele.instance.obtenirEtatEdition().currentUserControlPointChanged(uuid: noeud.obtenirNoeudOppose().obtenirUUID(),
                                                                                      pos: noeud.obtenirNoeudOppose().position)
            
            // TODO : À vérifier si on change la géométrie de la table ici ou non
            let table = noeud.parent as! NoeudTable
            table.updateGeometry()
        }
    }
    
    /// Visiter un mur pour le déplacement
    func visiterMur(noeud: NoeudMur) {
        if (noeud.estSelectionne()) {
            // Sauvegarder scale et rotation
            // TODO : trouver une meilleure solution pour conserver le scaling et la rotation lors d'un déplacement
            let scale = noeud.scale
            let rotation = noeud.rotation
            
            self.deplacerNoeud(noeud: noeud)
            
            // Remettre le scale et la rotation du mur
            noeud.scale = scale
            noeud.rotation = rotation
        }
    }
    
    /// Visiter un portail pour le déplacement
    func visiterPortail(noeud: NoeudPortail) {
        if (noeud.estSelectionne()) {
            self.deplacerNoeud(noeud: noeud)
        }
    }
    
    //virtual void visiterRondelle(NoeudRondelle* noeud);
    
    /// Déplace le noeud
    private func deplacerNoeud(noeud: NoeudCommun) {
        var pos = noeud.obtenirPositionRelative()
        pos = GLKVector3Add(pos, self.delta!)
        noeud.deplacer(position: pos)
        
        // Envoyer la commande
        FacadeModele.instance.obtenirEtatEdition().currentUserObjectTransformChanged(uuid: noeud.obtenirUUID(),
                                                                                     pos: noeud.position, rotation: MathHelper.determinerAngleAxeY(rotation: noeud.rotation), scale: noeud.scale)
    }
    
}

///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////

File: Livrable final/Sources/DLL/Arbre/Visiteurs/VisiteurDeplacement.cpp
///////////////////////////////////////////////////////////////////////////////
/// @file VisiteurDeplacement.cpp
/// @author Julien Charbonneau
/// @date 2016-09-17
/// @version 0.1
///
/// @addtogroup inf2990 INF2990
/// @{
///////////////////////////////////////////////////////////////////////////////

#include "VisiteurDeplacement.h"
#include "ModeleEtatJeu.h"
#include "ModeleEtatDeplacement.h"
#include "ModeleEtatPointControl.h"
#include "ModeleEtatDuplication.h"

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDeplacement::VisiteurDeplacement(glm::vec3 delta)
///
/// Constructeur de VisiteurDeplacement.
///
///	@param[in] delta: Vecteur trois dimensions pour le changement de position de l'objet
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
VisiteurDeplacement::VisiteurDeplacement(glm::vec3 delta, bool sendToServer ) {
	delta_ = delta;
	sendToServer_ = sendToServer;
}


////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDeplacement::VisiteurDeplacement(glm::vec3 delta)
///
/// Destructeur de VisiteurDeplacement.
///
///	@param[in] Aucun
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
VisiteurDeplacement::~VisiteurDeplacement() {

}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDeplacement::visiterAccelerateur(NoeudAccelerateur* noeud)
///
/// Cette fonction permet de visiter un accélérateur pour le déplacement.
///
///	@param[in] noeud: Noeud de l'objet
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurDeplacement::visiterAccelerateur(NoeudAccelerateur* noeud) {
	defaultVisit(noeud);
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDeplacement::visiterPointControl(NoeudPointControl * noeud)
///
/// Cette fonction permet de visiter un point de contrôle pour le déplacement.
///
///	@param[in] noeud: Noeud de l'objet
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurDeplacement::visiterPointControl(NoeudPointControl * noeud)
{
	if (noeud->estSelectionne()) {
		glm::dvec3 pos(noeud->obtenirPositionRelative());
		pos += delta_;

		// Deplacer le premier noeud
		noeud->deplacer(pos);

		// Deplacer son opposé
		noeud->obtenirNoeudOppose()->deplacer(pos * glm::dvec3(noeud->obtenirSymmetrie()));
		selectedNodes_.push_back(noeud);
	}
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDeplacement::visiterMur(NoeudMur* noeud)
///
/// Cette fonction permet de visiter un mur pour le déplacement.
///
///	@param[in] noeud: Noeud de l'objet
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurDeplacement::visiterMur(NoeudMur* noeud) {
	defaultVisit(noeud);
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurDeplacement::visiterPortail(NoeudPortail* noeud)
///
/// Cette fonction permet de visiter un portail pour le déplacement.
///
///	@param[in] noeud: Noeud de l'objet
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurDeplacement::visiterPortail(NoeudPortail* noeud) {
	defaultVisit(noeud);
}

void VisiteurDeplacement::visiterRondelle(NoeudRondelle * noeud)
{
}

void VisiteurDeplacement::defaultVisit(NoeudAbstrait* noeud)
{
	if (noeud->estSelectionne())
	{
		selectedNodes_.push_back(noeud);
		noeud->deplacer(noeud->obtenirPositionRelative() + delta_);
	
	}
}
///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////